### PR TITLE
Add extends_type argument to new_vctr()

### PR DESCRIPTION
**Pull request by hadley**
Closes #593

Currently failing due to #497

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -1,6 +1,13 @@
 
 # vctrs 0.2.0.9000
 
+* `new_vctr()` gains a `extends_type` argument which determines whether or not
+  the class of the underlying type will be include in the class. This generally
```
Comment (lionel-) [inline NEWS.md]:
```suggestion
  the class of the underlying type will be included in the class. This generally
```

**Comment by hadley** [comment]
That's fair. 

At some point we'll need to consider if this should use the implicit class. i.e. should `new_vctr(1)` get class numeric as well as double?

File: R/type-vctr.R [hunk 2]
```diff
@@ -49,16 +49,21 @@
...
-new_vctr <- function(.data, ..., class = character()) {
+new_vctr <- function(.data, ..., class = character(), inherit_base_type = TRUE) {
   if (!is_vector(.data)) {
     abort("`.data` must be a vector type.")
   }
 
   nms <- validate_names(.data)
-  attrib <- list(names = nms, ..., class = c(class, "vctrs_vctr"))
+
+  class <- c(class, "vctrs_vctr", if (inherit_base_type) class(.data))
```
Comment (DavisVaughan) [inline R/type-vctr.R]: Would it make sense for it to be `typeof(.data)` so we don't drag all of the `.data` classes along with it? Ideally I guess the caller ensured it was a bare vector with no classes. Also, there is no `typeof()` result for `"data.frame"`. But the others would work.

Comment (hadley) [reply R/type-vctr.R]: Oh yeah, I forgot that people might be feeding this things other than base classes.

Comment (hadley) [reply R/type-vctr.R]: This also revealed a subtle bug because `class(1)` is numeric, not double.

**Comment by DavisVaughan** [comment]
Potential alternative:

```
new_vctr(.data, ..., class = character(), super = typeof(.data)) {
  if (!is_vector(.data)) {
    abort("`.data` must be a vector type.")
  }
  
  nms <- validate_names(.data)
  
  class <- c(class, "vctrs_vctr", super)
  attrib <- list(names = nms, ..., class = class)
  
  vec_set_attributes(.data, attrib)
}
```

Would allow `super = NULL` to opt out, and `super = "data.frame"` if the user wanted that.

**Comment by lionel-** [comment]
I think we shouldn't allow arbitrary super classes, it should generally be considered UB.

**Pull request merged**